src/EVM/SymExec.hs

```diff
@@ -251,68 +264,105 @@ interpret fetcher maxIter askSmtIters vm =
-            -- consulting our SMT solver.
-            if iteration < fromMaybe 5 askSmtIters then do
-              let (r, vm') = runState (continue EVM.Unknown) vm
-              interpret fetcher maxIter askSmtIters vm' (k r)
-            else performQuery
+          PleaseAskSMT cond _ continue -> do
+            case cond of
+              -- is the condition concrete?
+              Lit c ->
+                -- have we reached max iterations, are we inside a loop?
+                case (maxIterationsReached vm maxIter, isLoopHead heuristic vm) of
+                  -- Yes. return a partial leaf
+                  (Just _, Just True) ->
+                    pure $ Partial vm.keccakEqs $ MaxIterationsReached vm.state.pc vm.state.contract
+                  -- No. keep executing
+                  _ ->
+                    let (r, vm') = runState (continue (Case (c > 0))) vm
+                    in interpret fetcher maxIter askSmtIters heuristic vm' (k r)
+
+              -- the condition is symbolic
+              _ ->
+                -- are in we a loop, have we hit maxIters, have we hit askSmtIters?
+                case (isLoopHead heuristic vm, askSmtItersReached vm askSmtIters, maxIterationsReached vm maxIter) of
+                  -- we're in a loop and maxIters has been reached
+                  (Just True, _, Just n) -> do
+                    -- continue execution down the opposite branch than the one that
+                    -- got us to this point and return a partial leaf for the other side
+                    let (r, vm') = runState (continue (Case $ not n)) vm
+                    a <- interpret fetcher maxIter askSmtIters heuristic vm' (k r)
+                    pure $ ITE cond a (Partial vm.keccakEqs (MaxIterationsReached vm.state.pc vm.state.contract))
+                  -- we're in a loop and askSmtIters has been reached
+                  (Just True, True, _) ->
+                    -- ask the smt solver about the loop condition
+                    performQuery
+                  -- otherwise just try both branches and don't ask the solver
+                  _ ->
+                    let (r, vm') = runState (continue EVM.Types.Unknown) vm
+                    in interpret fetcher maxIter askSmtIters heuristic vm' (k r)
 
           _ -> performQuery
 
       Stepper.EVM m -> do
         let (r, vm') = runState m vm
-        interpret fetcher maxIter askSmtIters vm' (k r)
+        interpret fetcher maxIter askSmtIters heuristic vm' (k r)
 
 maxIterationsReached :: VM -> Maybe Integer -> Maybe Bool
 maxIterationsReached _ Nothing = Nothing
 maxIterationsReached vm (Just maxIter) =
   let codelocation = getCodeLocation vm
-      iters = view (at codelocation % non 0) vm.iterations
+      (iters, _) = view (at codelocation % non (0, [])) vm.iterations
   in if num maxIter <= iters
      then Map.lookup (codelocation, iters - 1) vm.cache.path
      else Nothing
```

Comment:
Maybe trowing an error here would be more appropriate since we always expect to find the code location in the map and `Nothing` means that max iterations has been reached

Reply:
agree that it would be a bit nicer semantically, but I think it would make some other logic a little more complex, so I would probably prefer to just leave it as is for now.

The issue is that `iterations` is only set for a given codelocation in the callback executed after the `PleaseAskSMT` query is handled, so the first time we visit a JUMPI the iterations map can be empty. We could rework the logic that updates the iterations a bit, but honestly I don't think it would really be cleaner than what we have now...